sql/rowexec/fulltext_filter.go

```diff
@@ -0,0 +1,270 @@
+// Copyright 2023 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rowexec
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/expression"
+	"github.com/dolthub/go-mysql-server/sql/fulltext"
+	"github.com/dolthub/go-mysql-server/sql/plan"
+)
+
+// FulltextFilterTable handles row iteration for filters involving Full-Text indexes, as they behave differently than
+// other indexes. This acts as a sort of wrapper, so that integrators do not need to implement special logic on their
+// side.
+type FulltextFilterTable struct {
+	MatchAgainst *expression.MatchAgainst
+	Table        *plan.ResolvedTable
+}
+
+var _ sql.IndexedTable = (*FulltextFilterTable)(nil)
+
+// Name implements the interface sql.IndexedTable.
+func (f *FulltextFilterTable) Name() string {
+	return f.Table.Name()
+}
+
+// String implements the interface sql.IndexedTable.
+func (f *FulltextFilterTable) String() string {
+	return f.Table.String()
+}
+
+// Schema implements the interface sql.IndexedTable.
+func (f *FulltextFilterTable) Schema() sql.Schema {
+	return f.Table.Schema()
+}
+
+// Collation implements the interface sql.IndexedTable.
+func (f *FulltextFilterTable) Collation() sql.CollationID {
+	return f.Table.Collation()
+}
+
+// Partitions implements the interface sql.IndexedTable.
+func (f *FulltextFilterTable) Partitions(ctx *sql.Context) (sql.PartitionIter, error) {
+	if f.MatchAgainst.KeyCols.Type == fulltext.KeyType_None {
+		return f.Table.Partitions(ctx)
+	}
+	return &fulltextFilterTablePartitionIter{false}, nil
+}
+
+// PartitionRows implements the interface sql.IndexedTable.
+func (f *FulltextFilterTable) PartitionRows(ctx *sql.Context, partition sql.Partition) (sql.RowIter, error) {
+	// Keyless just iterates over the entire table. Not the most performant, but it works.
+	if f.MatchAgainst.KeyCols.Type == fulltext.KeyType_None {
+		return f.Table.PartitionRows(ctx, partition)
+	}
+
+	// Get the word parser for our string literal
+	words, err := f.MatchAgainst.Expr.Eval(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+	wordsStr, ok := words.(string)
+	if !ok {
+		return nil, fmt.Errorf("expected WORD to be a string, but had type `%T`", words)
+	}
+	collation := fulltext.GetCollationFromSchema(ctx, f.MatchAgainst.DocCountTable.Schema())
+	parser, err := fulltext.NewDefaultParser(ctx, collation, wordsStr)
```

Comment:
It's a bit tricky to explain. The index returns all rows that contain the words given in the match expression. The MatchAgainst expression then takes those rows and returns a relevancy value. That value is then converted to a boolean by the filter, which ultimately determines whether a row is returned. Without the index, the expression will still filter out every row, it will just process many unnecessary rows. This is also why the MatchAgainst expression works as an output, since it only cares about the row it was given.